invoice/src/main/java/org/killbill/billing/invoice/generator/InvoiceWithMetadata.java

```diff
@@ -69,11 +86,63 @@ private void build() {
                 tmp.resetNextRecurringDate();
             }
         }
-        if (invoice != null && invoice.getInvoiceItems().isEmpty()) {
-            invoice = null;
+
+
+        if (invoice != null ) {
+            // Filter $0 USAGE items if specified by config
+            if (filterZeroUsageItems) {
+                final Iterable<InvoiceItem> resultingItems = Iterables.filter(invoice.getInvoiceItems(), new Predicate<InvoiceItem>() {
+                    @Override
+                    public boolean apply(final InvoiceItem invoiceItem) {
+                        return invoiceItem.getInvoiceItemType() != InvoiceItemType.USAGE ||
+                               invoiceItem.getAmount().compareTo(BigDecimal.ZERO) != 0;
+                    }
+                });
+                final ImmutableList<InvoiceItem> filteredItems = ImmutableList.copyOf(resultingItems);
+                // Reset invoice items with filtered list
+                invoice.getInvoiceItems().clear();
+                invoice.addInvoiceItems(filteredItems);
+            }
+
+            // If no resulting items -> result a null invoice
+            // (but we already computed all the stuff we care about: chargedThroughDates, trackingIds,..
+            if (Iterables.isEmpty(invoice.getInvoiceItems())) {
+                invoice = null;
+            }
         }
     }
 
+
+    public static  final Map<UUID, DateTime> computeChargedThroughDates(final Invoice invoice, final InternalCallContext context)  {
+        final Map<UUID, DateTime> chargedThroughDates;
+        if (invoice != null &&
+            invoice.getStatus() == InvoiceStatus.COMMITTED) /* See https://github.com/killbill/killbill/issues/1296 */ {
+            chargedThroughDates = new HashMap<>();
+                for (final InvoiceItem item : invoice.getInvoiceItems()) {
```

Comment:
Maybe keep the original comment:

```
// Don't use invoice.getInvoiceItems(final Class<T> clazz) as some items can come from plugins
```

(it's an API we should probably remove at some point)